core/network/src/consensus_gossip.rs

```diff
@@ -91,15 +93,25 @@ impl<B: BlockT> ConsensusGossip<B> where B::Header: HeaderT<Number=u64> {
 			}
 			self.peers.insert(who, PeerConsensus {
 				known_messages,
+				is_authority: true,
+			});
+		}
+		else if roles.intersects(Roles::FULL) {
+			self.peers.insert(who, PeerConsensus {
+				known_messages: HashSet::new(),
+				is_authority: false,
 			});
 		}
 	}
 
 	fn propagate(&mut self, protocol: &mut Context<B>, message: message::Message<B>, hash: B::Hash) {
+		let len = self.peers.len() as f64;
 		for (id, ref mut peer) in self.peers.iter_mut() {
-			if peer.known_messages.insert(hash.clone()) {
-				trace!(target:"gossip", "Propagating to {}: {:?}", id, message);
-				protocol.send_message(*id, message.clone());
+			if peer.is_authority || rand::random::<f64>() < (len.sqrt() / len) {
```

Comment:
rather high variance. at least a sane minimum number of peers that get the message should be enforced (e.g. 3).

i think it should really be a randomly chosen `len.sqrt().max(3)` peers, as well as an validators that were not already in that random subset.